impl/src/main/java/org/glassfish/soteria/authorization/spi/impl/SubjectParser.java

```diff
@@ -389,181 +383,200 @@ private boolean tryWebLogic(String contextID, Collection<String> allDeclaredRole
-             
-                // EJBs return empty/null name for unauthenticated callers
-                return principalName != null && !principalName.isEmpty() ? ejbContext.getCallerPrincipal() : null;
+                // The EJB returned value must be verified for its "authenticated name" since it's vendor specific
+                return getVendorCallerPrincipal(ejbContext.getCallerPrincipal());
             }
 
             return null;
         }
-        
+
+        return null;
+    }
+
+    /**
+     * Get the underlying caller principal based on vendor specific (e.g.: class
+     * names, EJB unauthenticated name, etc)
+     *
+     * @param principal
+     * @return
+     */
+    private Principal getVendorCallerPrincipal(Principal principal) {
+        switch (principal.getClass().getName()) {
+            case "org.glassfish.security.common.PrincipalImpl": // GlassFish/Payara
+                return getPrincipalOrNull(principal, "ANONYMOUS");
+            case "com.ibm.ws.security.authentication.principals.WSPrincipal": // Liberty
+                return getPrincipalOrNull(principal, "UNAUTHENTICATED");
+            // JBoss EAP/WildFly convention 1 - single top level principal of the below type
+            case "org.jboss.security.SimplePrincipal":
+                return getPrincipalOrNull(principal, "anonymous");
+            // JBoss EAP/WildFly convention 2 - the one and only principal in group called CallerPrincipal
+            case "org.jboss.security.SimpleGroup":
+                if (principal.getName().equals("CallerPrincipal") && principal instanceof Group) {
+
+                    Enumeration<? extends Principal> groupMembers = ((Group) principal).members();
+
+                    if (groupMembers.hasMoreElements()) {
+                        return getPrincipalOrNull(groupMembers.nextElement(), "anonymous");
+                    }
+                }
+                break;
+            case "org.apache.tomee.catalina.TomcatSecurityService$TomcatUser": // TomEE
+                try {
+                    getPrincipalOrNull(
+                            (Principal) Class.forName("org.apache.catalina.realm.GenericPrincipal")
+                                    .getMethod("getUserPrincipal")
+                                    .invoke(
+                                            Class.forName("org.apache.tomee.catalina.TomcatSecurityService$TomcatUser")
+                                                    .getMethod("getTomcatPrincipal")
+                                                    .invoke(principal)), "guest");
+                } catch (Exception e) {
+
+                }
+                break;
+        }
+
         return null;
     }
-     
+
+    private Principal getPrincipalOrNull(Principal principal, String anonymousCallerName) {
+        EJBContext ejbContext = EJB.getEJBContext();
+        return !(ejbContext == null && anonymousCallerName.equals(principal.getName())) ? principal : null;
```

Comment:
I think this logic is the opposite of what you want. This will return the principal if `!(ejbContext == null)` -- i.e., any time you're in EJB context. What you really want is something more like this:
```
// if you're not in ejb context, or if you are, but the principal name doesn't match "unauthenticated", return the principal
(ejbContext == null || !anonymousCallerName.equals(principal.getName())) ? principal : null;
```